aten/src/ATen/native/Copy.cpp

```diff
@@ -124,21 +124,26 @@ static Tensor & copy_impl(Tensor & self, const Tensor & src, bool non_blocking)
   // 1. Memory Format for source and destination tensors is contiguous.
   // 2. Device for both the source and destination tensor is CPU.
   // 3. dtype conversion between FP32->FP16 and FP16->FP32.
+  // To avoid out of bounds memory access, this also checks whether there's
+  // enough self.numel() storage before copying src.numel() items, see
+  // fbgemm::Float16ToFloat_ref.
+  // https://github.com/pytorch/pytorch/issues/88543
   #ifdef USE_FBGEMM
     if (((self.dtype() == at::kFloat && src.dtype() == at::kHalf) ||
          (self.dtype() == at::kHalf && src.dtype() == at::kFloat)) &&
         (self.device().is_cpu() && src.device().is_cpu()) &&
         ((self.is_contiguous() && src.is_contiguous()) ||
-         (self.is_non_overlapping_and_dense() && self.strides() == src.strides()))) {
+         (self.is_non_overlapping_and_dense() && self.strides() == src.strides())) &&
+        (self.numel() >= src.numel())) {
```

Comment:
I check that we have enough elements to write to in `self` since `fbgemm::Float16ToFloat_ref` uses `numel`. Then I only read/write `src.numel` elements. Before, we used to read/write `self.numel` elements, but `src` might have less.

Also, because the fbgemm code here doesn't support broadcasting, it may lead to (perhaps) surprising results:

```py
In [30]: torch.rand((1,2,3), dtype=torch.float16).copy_(torch.ones(1))
Out[30]:
tensor([[[1.0000, 0.5332, 0.8677],
         [0.8384, 0.8584, 0.1699]]], dtype=torch.float16)

In [31]: torch.rand((1,2,3), dtype=torch.float64).copy_(torch.ones(1))
Out[31]:
tensor([[[1., 1., 1.],
         [1., 1., 1.]]], dtype=torch.float64)

```

In the fbgemm case, we only write part of the value. The rest remains uninitialized.

Reply:
Alternatively, we can change the numel check to `==`. Then all data should be copied. The only downside(?) is that some broadcasting cases will be handled by TI and not fbgemm after this. Personally, I think `==` is better, but maybe I'm missing some context.

Reply:
Consistency between dtypes should be driving us here. I think using `==` + leaving a comment saying that this path does not support broadcasting should be the answer

Reply:
Then, we shouldn't even need to change `self.numel()` for `src.numel()` below.

Reply:
Done. Note: I had to add another check for shapes being the same to guard against (4, 5, 6), (6, 5, 4).